### PR TITLE
Preserve terminal observation in BraxAutoResetWrapper for truncation bootstrap

### LINKED ISSUE #305: BraxAutoResetWrapper discards terminal observation, breaking correct value bootstrapping on truncation
# `BraxAutoResetWrapper` discards terminal observation, breaking correct value bootstrapping on truncation

Hello everyone, I have searched through the existing and closed issues but I have found nothing regarding incorrect value bootstrapping on truncation, so here it is:

## Summary

`BraxAutoResetWrapper` (in `mujoco_playground/_src/wrapper.py`) overwrites `state.obs` with the reset observation on `done`, without preserving the pre-reset observation anywhere in `state.info`.  While this is perfect for automated-reset in vectorized environments, this makes it impossible for downstream RL algorithms to implement correct value bootstrapping on truncation.

## Why this matters for RL

Every actor-critic method needs to distinguish termination from truncation when computing TD targets / GAE:

- **Termination** (e.g. agent fell): the MDP ends. Target is `r_t` (no bootstrap).
- **Truncation** (e.g. time limit hit, task still alive): the episode is artificially cut. Target is `r_t + γ·V(s_{t+1})`, where `s_{t+1}` is the observation the agent *would have seen next* had the episode continued.

`wrap_for_brax_training` already exposes the termination/truncation distinction correctly via `state.info['truncation']` (set by brax's `EpisodeWrapper`). But to actually compute the truncation bootstrap `V(s_{t+1})`, the learner needs access to `s_{t+1}` — the terminal observation of the truncated trajectory. After `BraxAutoResetWrapper.step`, that observation is gone: `state.obs` has been replaced by the reset state, and `info` only holds reset-related bookkeeping (`_first_data`, `_first_obs`, `_rng`, `_done_count`).

In practice this means every PPO/SAC implementation built on Playground is silently bootstrapping on `V(s_reset)` instead of `V(s_{t+1})` at every truncation. The bias is small on short-horizon tasks with low γ but grows with episode length and γ, and is particularly pathological on locomotion tasks where `s_reset` is an upright pose very different from the late-episode state.

The same flaw exists in `brax.envs.wrappers.training.AutoResetWrapper`. Gymnasium solved this years ago by putting the terminal observation in `info["final_observation"]` on truncation, which is the convention most SB3/CleanRL/RLlib code expects.

## Proposed fix

In `BraxAutoResetWrapper.step`, before the `where_done` mask replaces `obs` and `data`, stash the pre-reset observation in info:

```python
# inside BraxAutoResetWrapper.step, after self.env.step(...) returns
next_info = state.info
next_info[f"{self._info_key}_final_obs"] = state.obs
# ...rest unchanged
```

This is does not break anything, and enables correct truncation handling for any downstream learner that reads it.

## Reproduction

```python
import jax, jax.numpy as jnp
from mujoco_playground import registry, wrapper

env = wrapper.wrap_for_brax_training(registry.load("HopperHop"), episode_length=10)
state = jax.jit(env.reset)(jax.random.split(jax.random.PRNGKey(0), 2))
for t in range(10):
    state = jax.jit(env.step)(state, jnp.zeros((2, env.action_size)))
print("done:", state.done, "truncation:", state.info["truncation"])
print("obs (overwritten by reset):", state.obs[0, :3])
# no info key holds the pre-reset observation
print("info keys:", [k for k in state.info if "obs" in k.lower()])
```

Expected: some `info['...final_obs...']` key carrying the real terminal observation.
Actual: no such key exists.

Maybe I am missing something, please tell me if that's the case :)

Otherwise, I can take care of the pull-request if needed.

Yann

### PR DESCRIPTION
**Pull request by dparikh79**
## Summary

Fixes #305.

`BraxAutoResetWrapper.step` in `mujoco_playground/_src/wrapper.py` overwrites `state.obs` with the reset observation on `state.done` without preserving the pre-reset (terminal) observation. Downstream actor-critic learners therefore have no way to compute the correct value-bootstrap target on truncation:

```
target = r_t + gamma * V(s_{t+1})
```

where `s_{t+1}` is the observation the agent *would have seen next* had the episode not been truncated, not the reset observation. The result is that every PPO/SAC implementation built on Playground silently bootstraps on `V(s_reset)` at every truncation. The bias grows with episode length and gamma and is particularly bad on locomotion tasks where the reset pose is very different from the late-episode state. @YannBerthelot's issue write-up covers the math and the comparison to brax upstream / Gymnasium thoroughly.

## Change

`mujoco_playground/_src/wrapper.py`: stash `state.obs` (the pre-reset observation returned by the wrapped env's step) under `info['AutoResetWrapper_final_obs']` before the `where_done` mask replaces it. The new field is set on every step; learners can combine it with the existing `info['truncation']` signal to apply the correct bootstrap on truncation only.

This matches the convention already used by Gymnasium's `info['final_observation']` (since v0.26).

The fix is the same shape as the proposal in #305 and credits @YannBerthelot in the commit message.

## Test plan

Extended `WrapperTest.test_auto_reset_wrapper` in `mujoco_playground/_src/wrapper_test.py` so the existing `(full_reset, cache_reset)` parameterized cases additionally assert:

- [x] `info['AutoResetWrapper_final_obs']` is present after a done step.
- [x] Its shape matches `state.obs.shape`.
- [x] On a done step (where `state.obs` has been replaced by the reset observation), `final_obs` differs from `state.obs`; confirming the terminal observation, not the reset, was stashed.

- [ ] **Could not run pytest locally**: mujoco_playground's full install needs JAX + brax + MuJoCo + dm_control_suite. I am relying on CI to run the wrapper test plus the rest of `_src/` tests.

## CLA

I previously signed the Google CLA via dparikh79 for google-deepmind/mujoco, google-deepmind/mujoco_menagerie, and google-deepmind/dm_control. The umbrella should cover this contribution; if the CLA bot reports otherwise, happy to re-sign.

## AI Assistance Disclosure

Implementation and the regression test were drafted with Claude assistance, against @YannBerthelot's rigorous bug analysis in #305. I reviewed every changed line, traced the call into `BraxAutoResetWrapper.step` to confirm the where_done mask and `state.replace` semantics, and confirmed the test extends the existing parameterized assertions without breaking them. I am the human contributor accountable for this PR.

cc: @YannBerthelot

### CODE REVIEW

**Comment by btaba** [comment]
I missed #305 , that's a really good catch, thanks for applying the patch. This PR is not sufficient though, the final obs needs to be used downstream (say in brax). Are you planning to follow-up there? 

Additionally for the tests to pass, you need to add the `AutoResetWrapper_final_obs` key  on reset as well (see failing tests)

**Comment by dparikh79** [comment]
Thanks @btaba, pushed the fix in ad0fd5e: `reset()` now seeds `AutoResetWrapper_final_obs` with the initial observation so the info dict has the same key set on every transition (the JAX dict-key-mismatch was exactly that). The two failing `test_auto_reset_wrapper_full_reset` and `test_evalwrapper_with_reset_full_reset` should pass on the next CI run.

On the downstream brax-side follow-up: agreed, this PR on its own only stops the playground side from discarding the terminal observation. Whether downstream learners actually use it for the truncation bootstrap is on the learner side. Two paths from here, happy to go in whichever direction you prefer:

1. **Land this PR as-is** (just stops the loss of data on the playground side), and open a separate brax-side PR (or playground PPO/SAC trainer PR if one exists in this repo) to actually wire `info['AutoResetWrapper_final_obs'] + info['truncation']` into the value-bootstrap target.
2. **Bundle the downstream change into this PR** if you have a specific trainer in mind in playground that should be updated alongside.

If (1), I can start scoping a brax-side patch against `brax/training/agents/ppo/training.py`'s `compute_gae` (or wherever the truncation-bootstrap belongs in current brax). Let me know what you would prefer.
